File: 141-Linked-List-Cycle/solution02.py
#!/usr/bin/python
# ==============================================================================
# Author:   Tao Li (taoli@ucsd.edu)
# Date:     May 6, 2015
# Question: 141-Linked-List-Cycle
# Link:     https://leetcode.com/problems/linked-list-cycle/
# ==============================================================================
# Given a linked list, determine if it has a cycle in it.
# 
# Follow up:
# Can you solve it without using extra space?
# ==============================================================================
# Method: Fast & Slow Pointers
# Time Complexity: O(n)
# Space Complexity: O(1)
# ==============================================================================

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        if not head:
            return False
        p1 = p2 = head
        while p1 and p1.next:
            p1 = p1.next.next
            p2 = p2.next
            if id(p1) == id(p2):
                return True
        return False

if __name__ == '__main__':
    head = ListNode(0)
    head.next = head
    # head.next = ListNode(1)
    # head.next.next = ListNode(2)
    print Solution().hasCycle(head)

File: 125-Valid-Palindrome/Java/Solution02.java
// ==============================================================================
// Author:   Tao Li (taoli@ucsd.edu)
// Date:     Jul 31, 2015
// Question: 125-Valid-Palindrome
// Link:     https://leetcode.com/problems/valid-palindrome/
// ==============================================================================
// Given a string, determine if it is a palindrome, considering only 
// alphanumeric characters and ignoring cases.

// For example,
// "A man, a plan, a canal: Panama" is a palindrome.
// "race a car" is not a palindrome.

// Note:
// Have you consider that the string might be empty? This is a good question 
// to ask during an interview.

// For the purpose of this problem, we define empty string as valid palindrome.
// ==============================================================================
// Method: One pass
// Time Complexity: O(n)
// Space Complexity: O(1)
// ==============================================================================

import java.lang.*;
import java.util.*;

public class Solution02 {
    public static boolean isPalindrome(String s) {
        int i = 0, j = s.length() - 1;
        while (i < j) {
            while (i < j && !Character.isLetterOrDigit(s.charAt(i))) i ++;
            while (i < j && !Character.isLetterOrDigit(s.charAt(j))) j --;
            if (Character.toLowerCase(s.charAt(i)) != Character.toLowerCase(s.charAt(j))) return false;
            i ++; j --;
        }
        return true;
    }

    public static void main(String[] args) {
        String s = new String("121");
        System.out.println(isPalindrome(s));
    }
}